Trigger frame event on NMI, with trigger on vcounter() == 241 as fallback

Brunnis opened the pull request:
This is a fix that removes one full frame of input lag. There is already a pull request for a similar fix, but that fix was thought to have side effects in certain corner cases (enabling overscan mid-frame). See old pull request here:

https://github.com/libretro/bsnes-mercury/pull/17

This new code relies on the fact that the NMI (non-maskable interrupt) triggers right at the beginning of the VBLANK period. By exiting the emulator loop with a frame event right when the emulator raises NMI, we're able to exit before the input is read, thereby shaving off one frame of input lag.

The neat thing about this version of the fix is that it should work with games that modify the overscan bit mid-frame. Waiting for NMI should be reliable, since this is what game code itself does to detect when a frame has been rendered and when it's safe to run the game code. Please refer to the following link, section "SETINI - Screen Mode/Video Select" for overscan/NMI info:

http://wiki.superfamicom.org/snes/show/Registers

As you can see, the overscan bit's value directly affects when NMI occurs, which further strengthens the argument to use it as a frame event trigger.

As a safeguard, the original condition of exiting the emulator loop and reporting a frame event at V=241 is still there. If, for some reason, an NMI hasn't occurred when V=241, the behavior will be exactly the same as it is today.

**A note regarding bsnes-mercury vs higan v099:**

While comparing the source code of bsnes-mercury with the latest version of higan, I noticed some important discrepancies. While higan correctly detects overscan changes mid-frame by reading the register value directly, bsnes-mercury does not (at least not in all places). bsnes-mercury instead calls the PPU:overscan() function, which returns a value that was cached at the beginning of the frame. Example from /sfc/cpu/timing/irq.cpp:

**bsnes-mercury:**
`bool nmi_valid = (vcounter(2) >= (!ppu.overscan() ? 225 : 240));`

**higan v099**
`bool nmi_valid = vcounter(2) >= ppu.vdisp(); //Brunnis: vdisp() returns !regs.overscan ? 225 : 240`

Given this inaccuracy in bsnes-mercury regarding overscan handling, it would probably be okay to merge the original pull request. However, the new fix is better, since it doesn't have _any_ direct dependence on the overscan value.

Alcaro commented:
I tried looking for where the NMI interrupt is sent to the CPU, but couldn't find it. Looks like you found some other place to hook. Whatever, good enough.

> As a safeguard, the original condition of exiting the emulator loop and reporting a frame event at V=241 is still there.

You can disable NMI, so yes, good idea.

>\+  void scanline(bool * frame_event_performed);

Any reason that doesn't poke the status directly?

If yes, bool& feels cleaner than bool*.

>\+  if(!(*frame_event_performed) && cpu.vcounter() == 241)
>\+  {
>\+    scheduler.exit(Scheduler::ExitReason::FrameEvent);
>\+    *frame_event_performed = true;
>\+  }

The second condition is false nearly all the time, I believe the branch predictor is happier if you put it first. And there's no reason for that extra paren.

Finally, the snes9x cores have this under a #ifdef. Better do the same here, in case it does break. Or a core option, but that's a lot harder.

I'll merge this once I've figured out whether FrameEvent can affect any core-visible variables, other than cheats and input timing.

Brunnis commented:
> Any reason that doesn't poke the status directly?
Yes. I wanted to put the frame_event_performed in the status struct of the CPU. However, that struct is not available to System. Instead of complicating things, I simply pass a reference to the scanline() function.

> If yes, bool& feels cleaner than bool*.
Fixed.

> The second condition is false nearly all the time, I believe the branch predictor is happier if you put it first. And there's no reason for that extra paren.
Fixed.

> Finally, the snes9x cores have this under a #ifdef. Better do the same here, in case it does break. Or a core option, but that's a lot harder.
Fixed.

> In addition to my concerns in the post above, I must also require that someone verifies that the bottom scanline does not lag or otherwise glitch, both with overscan enabled and disabled.
I disabled the "Crop overscan" setting in RetroArch and ran tests on all three profiles. I used Super Mario World (no overscan) and Dragon Quest V (overscan). I then found a spot in both games where I could move vertically. I started moving vertically, paused emulation and then started frame-advancing while still moving vertically.

Result: Neither of the two games on any of the profiles showed any anomalies regarding the last scanline.

Brunnis commented:
Some more testing with games using overscan reported here: http://libretro.com/forums/showthread.php?t=5428&p=41937&viewfull=1#post41937

No issues found.
